pkg/apis/build/v1alpha1/build_validation.go

```diff
@@ -0,0 +1,182 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/knative/pkg/apis"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Validate Build
+func (b *Build) Validate() *apis.FieldError {
+	return validateObjectMetadata(b.GetObjectMeta()).ViaField("metadata").Also(b.Spec.Validate().ViaField("spec"))
+
+	// TODO: Builder Validation
+	// return ac.builder.Validate(b)
+	//	return nil
+}
+
+// Validate for build spec
+func (bs *BuildSpec) Validate() *apis.FieldError {
+	if bs.Template == nil && len(bs.Steps) == 0 {
+		return apis.ErrMissingField("b.spec.template").Also(apis.ErrMissingField("b.spec.steps"))
+		//validationError("NoTemplateOrSteps", "build must specify either template or steps")
+	}
+	if bs.Template != nil && len(bs.Steps) > 0 {
+		return apis.ErrMissingField("b.spec.template").Also(apis.ErrMissingField("b.spec.steps"))
+		//return validationError("TemplateAndSteps", "build cannot specify both template and steps")
+	}
+
+	if bs.Template != nil && bs.Template.Name == "" {
+		apis.ErrMissingField("build.spec.template.name")
+		//return validationError("MissingTemplateName", "template instantiation is missing template name: %v", b.Spec.Template)
+	}
+
+	if err := bs.ValidateSecrets(); err != nil {
+		return err
+	}
+	// If a build specifies a template, all the template's parameters without
+	// defaults must be satisfied by the build's parameters.
+	var volumes []corev1.Volume
+	//var tmpl BuildTemplateInterface
+	if bs.Template != nil {
+		tmplName := bs.Template.Name
+		if tmplName == "" {
+			return apis.ErrMissingField("build.spec.template.name")
+			//validationError("MissingTemplateName", "the build specifies a template without a name")
+		}
+		if bs.Template.Kind != "" {
+			return bs.Template.Validate()
+		}
+	}
+	if err := validateVolumes(append(bs.Volumes, volumes...)); err != nil {
+		return err
+	}
+
+	if err := validateSteps(bs.Steps); err != nil {
+		return err
+	}
+
+	// Need kubeclient to fetch template
```

Comment:
Yeah I think the current thinking is that we'll only do simple validation in the webhook, and leave it up to the controller to do full validation with resolved objects like the template, service account, secrets, etc.

I'm not sure there's a good story yet for surfacing validation errors outside of webhook checks though, we'd probably need to have some new condition type to express the `Invalid` state.